### ISSUE TITLE
SameDiff: Ops and Next Steps

### ISSUE THREAD

**Issue by AlexDBlack** [eclipse/deeplearning4j#7913]
Based an an analysis of all frozen model files I can find, here are the ops that we can't yet import:
```
AudioSpectrogram
BatchNormWithGlobalNormalization
CheckNumerics
DecodeJpeg
DecodeWav
Dequantize
FIFOQueueV2
FakeQuantWithMinMaxVars
Mfcc
NonMaxSuppressionV3
NonMaxSuppressionV4
PaddingFIFOQueue
QuantizeDownAndShrinkRange
QuantizeV2
QuantizedAvgPool
QuantizedBiasAdd
QuantizedConcat
QuantizedConv2D
QuantizedMatMul
QuantizedMaxPool
QuantizedRelu
QueueDequeue
QueueDequeueManyV2
TFLite_Detection_PostProcess
```

Full analysis: https://gist.github.com/AlexDBlack/2a389afde91d37e7da652b4f25185687

Note CheckNumerics and FakeQuantWithMinMaxVars should be supported but not mapped at Java level.
NonMaxSuppression and NonMaxSuppressionV2 can be imported; not V3/V4 (not sure how they differ, if at all).

Ops test coverage: (to be added later)

TF import ops coverage: (to be added later)

**Comment by raver119** [comment]
I've added CheckNumerics wrapper as well.

NonMaxSuppression will be investigated by @shugeo

**Comment by rnett** [comment]
From GPT-2, we're also missing `Multinomial` ([tensorflow](https://www.tensorflow.org/api_docs/python/tf/distributions/Multinomial)).

**Comment by AlexDBlack** [comment]
Updated list - test coverage and op mapping:
https://gist.github.com/AlexDBlack/47ff324a643dcc1d49ca871ac5ae3be9

Note coverage is an underestimate due to ignores (pending fixes). In a few cases, the TF op mapping name could be wrong too (for example HasInf -> IsInf).

**Comment by rnett** [comment]
Updated list counting tf import tests as foreward pass tests: [here](https://gist.github.com/rnett/6b31087dd715e56f682456c77287a644)

```
o.n.a.v.OpValidation - *****************************************************
o.n.a.v.OpValidation - Op Validation:                        182 of 404 classes with adequate tests (34.87% coverage)
o.n.a.v.OpValidation - Forward pass tests:                   268 of 404 classes (66.34% coverage)
o.n.a.v.OpValidation - Gradient check tests:                 215 of 424 classes (50.71% coverage)
o.n.a.v.OpValidation - (66 ops excluded from gradient check coverage)
o.n.a.v.OpValidation - (118 ops excluded from fwd+gradient tests)
o.n.a.v.OpValidation - TF mapped ops:                        543 of 1012 (53.66%)
o.n.a.v.OpValidation - SD ops with TF import mapping + test  204 of 290 (70.34%) - 13 ignored for coverage
o.n.a.v.OpValidation - Libnd4j mapped ops:                   321 of 434 (76.25%) - 13 excluded for coverage
o.n.a.v.OpValidation - *****************************************************
```

**Comment by AlexDBlack** [comment]
Multinomial test cases also need to be fixed - there's only the checkpoint, no frozen graph, no expected values.
Compare: https://github.com/deeplearning4j/dl4j-test-resources/tree/master/src/main/resources/tf_graphs/examples/multinomial/logits/sample/rank1
vs. https://github.com/deeplearning4j/dl4j-test-resources/tree/master/src/main/resources/tf_graphs/examples/meshgrid/n1_ij

cc @rnett

**Comment by rnett** [comment]
Should be fixed.

**Comment by agibsonccc** [comment]
Closing — SameDiff has been completely rewritten with a new DSP (Dynamic Shape Plan) execution engine, native plan compilation, CUDA graph replay, and Triton compilation. The architecture referenced in this issue no longer exists. If you still experience this issue with the upcoming release, please open a new issue.
